File: src/remote/privateDriveData.ts
import type { Child, Event, Parent } from "@/domain/models";

export interface KnownGroupEntry {
  groupId: string;
  groupName: string;
  sharedSheetId: string;
  role: "ADMIN" | "MEMBER";
}

export interface CachedRemoteIds {
  driveFileId: string | null;
  sheetIds: Record<string, string>;
}

/** Top-level shape of the JSON file stored in the parent's Drive. */
export interface PrivateDriveData {
  schemaVersion: number;
  parent: Parent;
  children: Child[];
  privateEvents: Event[];
  knownGroups: KnownGroupEntry[];
  cachedRemoteIds: CachedRemoteIds;
  updatedAt: number;
}

export function newPrivateDriveData(parent: Parent): PrivateDriveData {
  return {
    schemaVersion: 1,
    parent,
    children: [],
    privateEvents: [],
    knownGroups: [],
    cachedRemoteIds: { driveFileId: null, sheetIds: {} },
    updatedAt: 0,
  };
}

File: app/src/main/java/i/drive/kids/data/remote/PrivateDriveData.kt
package i.drive.kids.data.remote

import i.drive.kids.config.ParentPrivateConfig
import i.drive.kids.domain.model.Child
import i.drive.kids.domain.model.Event
import i.drive.kids.domain.model.Parent
import kotlinx.serialization.Serializable

/**
 * Root object stored as JSON in the parent's private Google Drive file
 * (`kids_rides_private_data.json`). Maps directly to the schema in DATA_MODEL.md §2.
 */
@Serializable
data class PrivateDriveData(
    val schemaVersion: Int = 1,
    val parent: Parent,
    val children: List<Child> = emptyList(),
    /** Events with visibilityScope == PRIVATE */
    val privateEvents: List<Event> = emptyList(),
    val preferences: ParentPrivateConfig = ParentPrivateConfig(),
    val knownGroups: List<KnownGroupEntry> = emptyList(),
    val cachedRemoteIds: CachedRemoteIds = CachedRemoteIds(),
    val updatedAt: Long = 0L
)

@Serializable
data class KnownGroupEntry(
    val groupId: String,
    val groupName: String,
    val sharedSheetId: String,
    val role: String = "MEMBER"
)

@Serializable
data class CachedRemoteIds(
    val driveFileId: String = "",
    /** Maps groupId -> sheetId */
    val sheetIds: Map<String, String> = emptyMap()
)
